### PR TITLE
semantic-python: compile early and terminating return statements

### PR DESCRIPTION
**Pull request by patrickt**
Because Python, unlike Ruby, has no implicit returns, it’s important for testing purposes to provide some sort of support for `return` statements. This patch augments the `Compile` typeclass to allow this.

Huge, huge 🎩 to @robrix, whose guidance as to how to implement early returns was really essential.

### CODE REVIEW

File: semantic-python/semantic-python.cabal [hunk 1]
```diff
@@ -64,6 +64,7 @@ test-suite test
                , directory ^>= 1.3.3
                , exceptions ^>= 0.10.2
                , filepath ^>= 1.4.2.1
+               , pretty-show ^>= 1.9.5
```
Comment (patrickt) [inline semantic-python/semantic-python.cabal]: Provides better output in test failures.

File: semantic-python/src/Language/Python/Core.hs [hunk 2]
```diff
@@ -1,4 +1,6 @@
-{-# LANGUAGE DefaultSignatures, DisambiguateRecordFields, FlexibleContexts, FlexibleInstances, OverloadedStrings, ScopedTypeVariables, NamedFieldPuns, TypeOperators #-}
+{-# LANGUAGE DefaultSignatures, DeriveAnyClass, DerivingStrategies, DerivingVia, DisambiguateRecordFields,
+             FlexibleContexts, FlexibleInstances, NamedFieldPuns, OverloadedStrings, ScopedTypeVariables,
+             StandaloneDeriving, TypeOperators, UndecidableInstances, DeriveGeneric #-}
```
Comment (patrickt) [inline semantic-python/src/Language/Python/Core.hs]: image

File: semantic-python/src/Language/Python/Core.hs [hunk 3]
```diff
@@ -16,13 +18,35 @@ import qualified TreeSitter.Python.AST as Py
...
 defaultCompile t = fail $ "compilation unimplemented for " <> show t
 
-instance (Compile l, Compile r) => Compile (Either l r) where compile = compileSum
+newtype CompileSum py = CompileSum py
+
+instance (Generic py, GCompileSum (Rep py)) => Compile (CompileSum py) where
+  compile (CompileSum a) = compileSum a
+  compileCC (CompileSum a) cc = compileCCSum a cc
+
+deriving via CompileSum (Either l r) instance (Compile l, Compile r) => Compile (Either l r)
```
Comment (patrickt) [inline semantic-python/src/Language/Python/Core.hs]: This is the best line of code in the patch.

File: semantic-python/src/Language/Python/Core.hs [hunk 4]
```diff
@@ -16,13 +18,35 @@ import qualified TreeSitter.Python.AST as Py
 class Compile py where
   -- FIXME: we should really try not to fail
   compile :: (Member Core sig, Carrier sig t, Foldable t, MonadFail m) => py -> m (t Name)
+
   default compile :: (MonadFail m, Show py) => py -> m (t Name)
   compile = defaultCompile
 
+  compileCC :: (Member Core sig, Carrier sig t, Foldable t, MonadFail m) => py -> m (t Name) -> m (t Name)
```
Comment (patrickt) [inline semantic-python/src/Language/Python/Core.hs]: This is like `compile`, but it takes a continuation—the operation to sequence next—as an argument. This means that we can encode the compilation of return statements, no matter if they’re early or terminal, by using `compileCC` and explicitly ignoring the continuation

File: semantic-python/src/Language/Python/Core.hs [hunk 5]
```diff
@@ -16,13 +18,35 @@ import qualified TreeSitter.Python.AST as Py
...
+
 defaultCompile :: (MonadFail m, Show py) => py -> m (t Name)
 defaultCompile t = fail $ "compilation unimplemented for " <> show t
 
-instance (Compile l, Compile r) => Compile (Either l r) where compile = compileSum
+newtype CompileSum py = CompileSum py
+
+instance (Generic py, GCompileSum (Rep py)) => Compile (CompileSum py) where
+  compile (CompileSum a) = compileSum a
+  compileCC (CompileSum a) cc = compileCCSum a cc
```
Comment (robrix) [inline semantic-python/src/Language/Python/Core.hs]: What do you think about removing `compileSum` & `compileCCSum` so we can’t accidentally use one but not the other?

File: semantic-python/src/Language/Python/Core.hs [hunk 6]
```diff
@@ -104,6 +135,12 @@ instance Compile Py.IfStatement where
           clause (Left  Py.ElifClause{ condition, consequence }) rest  =
             if' <$> compile condition <*> compile consequence <*> rest
 
+  compileCC Py.IfStatement{ condition, consequence, alternative} cc =
```
Comment (robrix) [inline semantic-python/src/Language/Python/Core.hs]: Can we define `compile` in terms of our new definition of `compileCC`? E.g. `compile p = compileCC p none`?

File: semantic-python/src/Language/Python/Core.hs [hunk 7]
```diff
@@ -1,6 +1,10 @@
+<<<<<<< HEAD
```
Comment (robrix) [inline semantic-python/src/Language/Python/Core.hs]: 😬

Comment (patrickt) [reply semantic-python/src/Language/Python/Core.hs]: Oopsie.

**Pull request merged**